infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesPersistentVolumeClaims.java

```diff
@@ -231,10 +206,70 @@ public void onClose(KubernetesClientException cause) {
       }
     } catch (KubernetesClientException e) {
       throw new KubernetesInfrastructureException(e);
-    } finally {
-      if (watch != null) {
-        watch.close();
-      }
     }
   }
+
+  private Watch pvcIsBoundWatcher(
+      CompletableFuture<PersistentVolumeClaim> future,
+      Resource<PersistentVolumeClaim, DoneablePersistentVolumeClaim> pvcResource) {
+    return pvcResource.watch(
+        new Watcher<PersistentVolumeClaim>() {
+          @Override
+          public void eventReceived(Action action, PersistentVolumeClaim pvc) {
+            if (pvc.getStatus().getPhase().equals(PVC_BOUND_PHASE)) {
+              LOG.debug("pvc '" + pvc.getMetadata().getName() + "' is bound");
+              future.complete(pvc);
+            }
+          }
+
+          @Override
+          public void onClose(KubernetesClientException cause) {
+            if (cause != null) {
+              future.completeExceptionally(
+                  new InfrastructureException(
+                      "Waiting for persistent volume claim '"
+                          + pvcResource.get().getMetadata().getName()
+                          + "' was interrupted"));
+            } else if (!future.isDone()) {
+              future.cancel(true);
+            }
+          }
+        });
+  }
+
+  /**
+   * Creates and returns {@link Watch} that watches for 'WaitForFirstConsumer' events on given PVC.
+   */
+  private Watch pvcIsWaitingForConsumerWatcher(
+      CompletableFuture<PersistentVolumeClaim> future, PersistentVolumeClaim actualPvc)
+      throws InfrastructureException {
+    return clientFactory
+        .create(workspaceId)
+        .events()
+        .inNamespace(namespace)
+        .withField(PVC_EVENT_REASON_FIELD_KEY, PVC_EVENT_WAIT_CONSUMER_REASON)
+        .withField(PVC_EVENT_UID_FIELD_KEY, actualPvc.getMetadata().getUid())
+        .watch(
+            new Watcher<Event>() {
+              @Override
+              public void eventReceived(Action action, Event resource) {
+                LOG.debug(
+                    "PVC '"
+                        + actualPvc.getMetadata().getName()
+                        + "' is waiting for first consumer. Don't wait to bound to avoid deadlock.");
+                future.complete(actualPvc);
+              }
+
+              @Override
+              public void onClose(KubernetesClientException cause) {
```

Comment:
Should the cancel logic in `pvcIsBoundWatcher` also be included here?

Reply:
yes, fixed